### PR TITLE
nixos/gonic: init

### LINKED ISSUE #181965: gonic NixOS module
**Project description**
music streaming server / subsonic server API implementation

**Metadata**

* homepage URL: https://github.com/sentriz/gonic
* source URL: https://github.com/sentriz/gonic
* license: gpl3
* platforms: linux
* package: https://github.com/NixOS/nixpkgs/blob/master/pkgs/servers/gonic/default.nix

The package is already there, maintained by @Profpatsch. It'd be nice to have a module too!

### PR DESCRIPTION
**Pull request by autrimpo**
###### Description of changes

Adds a module for `gonic` using a generic `settings` style interface, including a basic module test. Based mainly on the `navidrome` module. Closes #181965.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
  - [x] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/services/audio/gonic.nix [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+        RootDirectory = "/run/gonic";
+        ReadWritePaths = "";
+        BindReadOnlyPaths = [
+          # gonic can access scrobbling services
+          "${config.environment.etc."ssl/certs/ca-certificates.crt".source}:/etc/ssl/certs/ca-certificates.crt"
+          builtins.storeDir
+          cfg.settings.podcast-path
+        ] ++ cfg.settings.music-path
+        ++ lib.optional (cfg.settings ? tls-cert) cfg.settings.tls-cert
+        ++ lib.optional (cfg.settings ? tls-key) cfg.settings.tls-key;
```
Comment (SuperSandro2000) [inline nixos/modules/services/audio/gonic.nix:57]: Please define the setting instead and check if it is null or set.

File: nixos/modules/services/audio/gonic.nix [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+        DynamicUser = true;
+        StateDirectory = "gonic";
+        CacheDirectory = "gonic";
+        WorkingDirectory = "/var/lib/gonic";
+        RuntimeDirectory = "gonic";
+        RootDirectory = "/run/gonic";
+        ReadWritePaths = "";
+        BindReadOnlyPaths = [
+          # gonic can access scrobbling services
+          "${config.environment.etc."ssl/certs/ca-certificates.crt".source}:/etc/ssl/certs/ca-certificates.crt"
```
Comment (SuperSandro2000) [inline nixos/modules/services/audio/gonic.nix:52]:
```suggestion
          "-/etc/ssl/certs/ca-certificates.crt"
```
We assume this in to many places already.

File: nixos/modules/services/audio/gonic.nix [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+          builtins.storeDir
```
Comment (SuperSandro2000) [inline nixos/modules/services/audio/gonic.nix:53]: We shouldn't use the storeDir here. It literally means everything that is deployed on the machine minus the state.

Are there any alternatives? What files do you need?

File: nixos/modules/services/audio/gonic.nix [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+  config = mkIf cfg.enable {
+    systemd.services.gonic = {
+      description = "Gonic Media Server";
+      after = [ "network.target" ];
+      wantedBy = [ "multi-user.target" ];
+      serviceConfig = {
+        ExecStart = ''
+          ${pkgs.gonic}/bin/gonic -config-path ${settingsFormat.generate "gonic" cfg.settings}
+        '';
```
Comment (SuperSandro2000) [inline nixos/modules/services/audio/gonic.nix:42]:
```suggestion
        ExecStart = "${pkgs.gonic}/bin/gonic -config-path ${settingsFormat.generate "gonic" cfg.settings}";
```
Thats not a multiline setting.

File: nixos/modules/services/audio/gonic.nix [hunk 5]
```diff
@@ -0,0 +1,81 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.services.gonic;
+  settingsFormat = pkgs.formats.keyValue { mkKeyValue = lib.generators.mkKeyValueDefault {} " ";
+                                           listsAsDuplicateKeys = true; };
```
Comment (SuperSandro2000) [inline nixos/modules/services/audio/gonic.nix:8]:
```suggestion
  settingsFormat = pkgs.formats.keyValue {
    mkKeyValue = lib.generators.mkKeyValueDefault {} " ";
    listsAsDuplicateKeys = true;
  };
```

File: nixos/modules/services/audio/gonic.nix (hunk 2 above)
Comment (autrimpo) [reply nixos/modules/services/audio/gonic.nix]: I assume the `-` at the beginning shouldn't be there so I left that out.

File: nixos/modules/services/audio/gonic.nix (hunk 3 above)
Comment (autrimpo) [reply nixos/modules/services/audio/gonic.nix:55]: ffmpeg for transcoding and mpv for jukebox. Both are inputs to the package, can we get them from that somehow?

**Comment by autrimpo** [comment]
I've noticed that scrobbling is broken so I marked it as a draft for now.

File: nixos/modules/services/audio/gonic.nix (hunk 1 above)
Comment (autrimpo) [reply nixos/modules/services/audio/gonic.nix:58]: I've reverted this change; both null and empty string values create entries in the final config which we don't want.

**Comment by autrimpo** [comment]
Scrobbling was fixed by adding `/etc/resolv.conf` to the list of read-only binds.

**Comment by autrimpo** [comment]
@SuperSandro2000 could I kindly ask you for another review? Thanks!

File: nixos/modules/services/audio/gonic.nix (hunk 3 above)
Comment (SuperSandro2000) [reply nixos/modules/services/audio/gonic.nix:54]: You could add them to passthru of the package to access them here.

File: nixos/modules/services/audio/gonic.nix (hunk 1 above)
Comment (SuperSandro2000) [reply nixos/modules/services/audio/gonic.nix:58]: We could check if the value is null.

**Comment by autrimpo** [comment]
Rebased to the current master

File: nixos/modules/services/audio/gonic.nix (hunk 3 above)
Comment (autrimpo) [reply nixos/modules/services/audio/gonic.nix]: done

File: nixos/modules/services/audio/gonic.nix (hunk 1 above)
Comment (autrimpo) [reply nixos/modules/services/audio/gonic.nix]: I think I found a solution that allows checking for null here - I do a null check and conditionally remove attributes when generating the final config

**Comment by autrimpo** [comment]
Rebased to current master

File: pkgs/servers/gonic/default.nix [hunk 6]
```diff
@@ -40,6 +41,11 @@ buildGoModule rec {
         '"${lib.getBin mpv}/bin/mpv"'
   '';
 
+  passthru = {
+    tests.gonic = nixosTests.gonic;
+    runtimeDeps = [ ffmpeg mpv ];
```
Comment (SuperSandro2000) [inline pkgs/servers/gonic/default.nix:46]:
```suggestion
```

File: nixos/modules/services/audio/gonic.nix [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+        CacheDirectory = "gonic";
+        WorkingDirectory = "/var/lib/gonic";
+        RuntimeDirectory = "gonic";
+        RootDirectory = "/run/gonic";
+        ReadWritePaths = "";
+        BindReadOnlyPaths = [
+          # gonic can access scrobbling services
+          "-/etc/ssl/certs/ca-certificates.crt"
+          cfg.settings.podcast-path
+        ] ++ pkgs.gonic.runtimeDeps
```
Comment (SuperSandro2000) [inline nixos/modules/services/audio/gonic.nix:62]:
```suggestion
        ] ++ pkgs.gonic
```
this should be enough

Comment (autrimpo) [reply nixos/modules/services/audio/gonic.nix:62]: has something changed with regards to https://github.com/NixOS/nixpkgs/pull/207106#discussion_r1162965931 ?

Comment (SuperSandro2000) [reply nixos/modules/services/audio/gonic.nix:62]: 🤔 I am actually not sure if this recursively works, it's just that this is somewhat unusual to see

Comment (autrimpo) [reply nixos/modules/services/audio/gonic.nix]: after some testing, I could not get it to work without including `builtins.storeDir` - I think we'd need to include a list of all transitive dependencies, including dynamically linked libraries

Comment (SuperSandro2000) [reply nixos/modules/services/audio/gonic.nix]: I just double checked and other services seem to do the same. This is probably one of those exceptions where using storeDir is good, normally you would want to use it rather rarely.

**Pull request merged**